File: samples/Lua/TestLua/Resources/luaScript/ZwoptexTest/ZwoptexTest.lua
local scheduler = CCDirector:sharedDirector():getScheduler()
--------------------------------------------------------------------
--
-- ZwoptexGenericTest
--
--------------------------------------------------------------------
local function ZwoptexGenericTest()
    local ret = createTestLayer("Zwoptex Tests",
                                "Coordinate Formats, Rotation, Trimming, flipX/Y")
    local spriteFrameIndex = 0
    local counter = 0
    local s = CCDirector:sharedDirector():getWinSize()
    local schedulerEntry = nil
    local schedulerFlipSpriteEntry = nil
    local sprite1 = nil
    local sprite2 = nil

    local function onEnter()
        CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("zwoptex/grossini.plist")
        CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("zwoptex/grossini-generic.plist")

        local layer1 = CCLayerColor:create(ccc4(255, 0, 0, 255), 85, 121)
        layer1:setPosition(ccp(s.width/2-80 - (85.0 * 0.5), s.height/2 - (121.0 * 0.5)))
        ret:addChild(layer1)

        sprite1 = CCSprite:createWithSpriteFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("grossini_dance_01.png"))
        sprite1:setPosition(ccp( s.width/2-80, s.height/2))
        ret:addChild(sprite1)

        sprite1:setFlipX(false)
        sprite1:setFlipY(false)

        local layer2 = CCLayerColor:create(ccc4(255, 0, 0, 255), 85, 121)
        layer2:setPosition(ccp(s.width/2+80 - (85.0 * 0.5), s.height/2 - (121.0 * 0.5)))
        ret:addChild(layer2)

        sprite2 = CCSprite:createWithSpriteFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName("grossini_dance_generic_01.png"))
        sprite2:setPosition(ccp( s.width/2 + 80, s.height/2))
        ret:addChild(sprite2)

        sprite2:setFlipX(false)
        sprite2:setFlipY(false)

        local function flipSprites(dt)
            counter = counter + 1

            local fx = false
            local fy = false
            local i = counter % 4

            if i == 0 then
                fx = false
                fy = false
            elseif i == 1 then
                fx = true
                fy = false
            elseif i == 2 then
                fx = false
                fy = true
            elseif i == 3 then
                fx = true
                fy = true
            end

            sprite1:setFlipX(fx)
            sprite2:setFlipX(fx)
            sprite1:setFlipY(fy)
            sprite2:setFlipY(fy)

            spriteFrameIndex = spriteFrameIndex + 1
            if spriteFrameIndex > 14 then
                spriteFrameIndex = 1
            end

            local str1 = string.format("grossini_dance_%02d.png", spriteFrameIndex)
            local str2 = string.format("grossini_dance_generic_%02d.png", spriteFrameIndex)
            sprite1:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(str1))
            sprite2:setDisplayFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(str2))
        end

        sprite1:retain()
        sprite2:retain()
        counter = 0
        local function startIn05Secs(dt)
            scheduler:unscheduleScriptEntry(schedulerEntry)
            schedulerFlipSpriteEntry = scheduler:scheduleScriptFunc(flipSprites, 0.5, false)
        end
        schedulerEntry = scheduler:scheduleScriptFunc(startIn05Secs, 1.0, false)
    end

    local function onExit()
        if schedulerEntry ~= nil then
            scheduler:unscheduleScriptEntry(schedulerEntry)
        end
        if schedulerFlipSpriteEntry ~= nil then
            scheduler:unscheduleScriptEntry(schedulerFlipSpriteEntry)
        end
        sprite1:release()
        sprite2:release()
        local cache = CCSpriteFrameCache:sharedSpriteFrameCache()
        cache:removeSpriteFramesFromFile("zwoptex/grossini.plist")
        cache:removeSpriteFramesFromFile("zwoptex/grossini-generic.plist")
    end

    local function onNodeEvent(event)
        if event == "enter" then
            onEnter()
        elseif event == "exit" then
            onExit()
        end
    end

    ret:registerScriptHandler(onNodeEvent)

    return ret
end

function ZwoptexTestMain()
    cclog("ZwoptexTestMain")
    Helper.index = 1
    local scene = CCScene:create()
    Helper.createFunctionTable = {
        ZwoptexGenericTest
    }
    scene:addChild(ZwoptexGenericTest())
    scene:addChild(CreateBackMenuItem())
    return scene
end

File: samples/Cpp/TestCpp/Classes/ZwoptexTest/ZwoptexTest.cpp
#include "ZwoptexTest.h"
#include "../testResource.h"

#define MAX_LAYER    1

static int sceneIdx = -1;

Layer* nextZwoptexTest();
Layer* backZwoptexTest();
Layer* restartZwoptexTest();

Layer* createZwoptexLayer(int nIndex)
{
    switch(nIndex)
    {
    case 0: return new ZwoptexGenericTest();
    }

    return NULL;
}

Layer* nextZwoptexTest()
{
    sceneIdx++;
    sceneIdx = sceneIdx % MAX_LAYER;

    Layer* pLayer = createZwoptexLayer(sceneIdx);
    pLayer->autorelease();

    return pLayer;
}

Layer* backZwoptexTest()
{
    sceneIdx--;
    int total = MAX_LAYER;
    if( sceneIdx < 0 )
        sceneIdx += total;    

    Layer* pLayer = createZwoptexLayer(sceneIdx);
    pLayer->autorelease();

    return pLayer;
}

Layer* restartZwoptexTest()
{
    Layer* pLayer = createZwoptexLayer(sceneIdx);
    pLayer->autorelease();

    return pLayer;
} 

//------------------------------------------------------------------
//
// ZwoptexTest
//
//------------------------------------------------------------------
void ZwoptexTest::onEnter()
{
    BaseTest::onEnter();
}

void ZwoptexTest::restartCallback(Object* pSender)
{
    Scene *s = ZwoptexTestScene::create();
    s->addChild(restartZwoptexTest());
    Director::sharedDirector()->replaceScene(s);
}

void ZwoptexTest::nextCallback(Object* pSender)
{
    Scene *s = ZwoptexTestScene::create();
    s->addChild(nextZwoptexTest());
    Director::sharedDirector()->replaceScene(s);
}

void ZwoptexTest::backCallback(Object* pSender)
{
    Scene *s = ZwoptexTestScene::create();
    s->addChild(backZwoptexTest());
    Director::sharedDirector()->replaceScene(s);
}

std::string ZwoptexTest::title()
{
    return "No title";
}

std::string ZwoptexTest::subtitle()
{
    return "";
}

//------------------------------------------------------------------
//
// ZwoptexGenericTest
//
//------------------------------------------------------------------
void ZwoptexGenericTest::onEnter()
{
    ZwoptexTest::onEnter();

    Size s = Director::sharedDirector()->getWinSize();

    SpriteFrameCache::sharedSpriteFrameCache()->addSpriteFramesWithFile("zwoptex/grossini.plist");
    SpriteFrameCache::sharedSpriteFrameCache()->addSpriteFramesWithFile("zwoptex/grossini-generic.plist");
    
    LayerColor *layer1 = LayerColor::create(ccc4(255, 0, 0, 255), 85, 121);
    layer1->setPosition(ccp(s.width/2-80 - (85.0f * 0.5f), s.height/2 - (121.0f * 0.5f)));
    addChild(layer1);

    sprite1 = Sprite::createWithSpriteFrame(SpriteFrameCache::sharedSpriteFrameCache()->spriteFrameByName("grossini_dance_01.png"));
    sprite1->setPosition(ccp( s.width/2-80, s.height/2));
    addChild(sprite1);

    sprite1->setFlipX(false);
    sprite1->setFlipY(false);

    LayerColor *layer2 = LayerColor::create(ccc4(255, 0, 0, 255), 85, 121);
    layer2->setPosition(ccp(s.width/2+80 - (85.0f * 0.5f), s.height/2 - (121.0f * 0.5f)));
    addChild(layer2);
    
    sprite2 = Sprite::createWithSpriteFrame(SpriteFrameCache::sharedSpriteFrameCache()->spriteFrameByName("grossini_dance_generic_01.png"));
    sprite2->setPosition(ccp( s.width/2 + 80, s.height/2));
    addChild(sprite2);

    sprite2->setFlipX(false);
    sprite2->setFlipY(false);

    schedule(schedule_selector(ZwoptexGenericTest::startIn05Secs), 1.0f);
    
    sprite1->retain();
    sprite2->retain();

    counter = 0;
}

void ZwoptexGenericTest::startIn05Secs(float dt)
{
    unschedule(schedule_selector(ZwoptexGenericTest::startIn05Secs));
    schedule(schedule_selector(ZwoptexGenericTest::flipSprites), 0.5f);
}

static int spriteFrameIndex = 0;
void ZwoptexGenericTest::flipSprites(float dt)
{
    counter++;

    bool fx = false;
    bool fy = false;
    int i = counter % 4;

    switch ( i )
    {
        case 0:
            fx = false;
            fy = false;
            break;
        case 1:
            fx = true;
            fy = false;
            break;
        case 2:
            fx = false;
            fy = true;
            break;
        case 3:
            fx = true;
            fy = true;
            break;
    }

    sprite1->setFlipX(fx);
    sprite2->setFlipX(fx);
    sprite1->setFlipY(fy);
    sprite2->setFlipY(fy);

    if(++spriteFrameIndex > 14)
    {
        spriteFrameIndex = 1;
    }

    char str1[32] = {0};
    char str2[32] = {0};
    sprintf(str1, "grossini_dance_%02d.png", spriteFrameIndex);
    sprintf(str2, "grossini_dance_generic_%02d.png", spriteFrameIndex);
    sprite1->setDisplayFrame(SpriteFrameCache::sharedSpriteFrameCache()->spriteFrameByName(str1));
    sprite2->setDisplayFrame(SpriteFrameCache::sharedSpriteFrameCache()->spriteFrameByName(str2));
}

ZwoptexGenericTest::~ZwoptexGenericTest()
{
    sprite1->release();
    sprite2->release();
    SpriteFrameCache *cache = SpriteFrameCache::sharedSpriteFrameCache();
    cache->removeSpriteFramesFromFile("zwoptex/grossini.plist");
    cache->removeSpriteFramesFromFile("zwoptex/grossini-generic.plist");
}

std::string ZwoptexGenericTest::title()
{
    return "Zwoptex Tests";
}

std::string ZwoptexGenericTest::subtitle()
{
    return "Coordinate Formats, Rotation, Trimming, flipX/Y";
}

void ZwoptexTestScene::runThisTest()
{
    Layer* pLayer = nextZwoptexTest();
    addChild(pLayer);

    Director::sharedDirector()->replaceScene(this);
}
